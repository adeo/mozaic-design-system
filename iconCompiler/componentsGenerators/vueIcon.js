const { parse } = require('svg-parser')
const { createComponentName, convertPxToRem } = require('../utils/tools')

const fileHeader = `/*
DON'T EDIT THIS FILE
It was auto-generated by "iconCompiler" script
*/

const GenerateIcon = (size, viewBox, childs) => ({
  props: {
    size: {
      type: String,
      default: size,
    },
    fill: {
      type: String,
      default: 'currentColor',
    },
  },
  render: function (createElement) {
    return createElement(
      attrs: {
        height: this.size,
        width: this.size,
        fill: this.fill,
        viewBox: viewBox
      },
      childs.map(elem => createElement(...elem))
    );
  },
});
`

const vueIconComponent = ({ fileName, data }) => {
  const parsedFileName = `MIcon${createComponentName(fileName)}`

  const sizePx = fileName
    .replace('.svg', '')
    .split('_')
    .pop()

  const size = convertPxToRem(sizePx.replace('px', ''))
  const parsedSVG = parse(data)
  const viewBox = parsedSVG.children[0].properties.viewBox

  const content = parsedSVG.children[0].children

  if (!size || !viewBox || !content || content.length === 0) {
    console.error(
      `âœ— ERROR : there was a problem building ${fileName} for vueJS`
    )

    return null
  }

  return `

export const ${parsedFileName} = GenerateIcon('${size}', '${viewBox}', [
  ${content.map(
    elem => `[
    '${elem.tagName}',
    { attrs: ${JSON.stringify(elem.properties)} },
  ]`
  )}
])`
}

const vueIconsIndex = iconList => `<script>
${fileHeader}
${iconList.map(vueIconComponent).join('')}

export default {}
</script>`

module.exports = vueIconsIndex
