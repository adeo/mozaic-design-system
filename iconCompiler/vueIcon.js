const { createComponentName, convertPxToRem } = require('./tools')

const fileHeader = `/*
DON'T EDIT THIS FILE
It was auto-generated by "iconCompiler" script
*/`

const vueIconComponent = ({ fileName, data, size }) => {
  const parsedFileName = createComponentName(fileName)
  const parsedData = data.replace(
    '<svg',
    '<svg height={this.size} width={this.size} fill={this.fill}'
  )
  return `
  export const ${parsedFileName} = {
    render() {
      return (
        ${parsedData}
      )
    },
    props: {
      size: {
        type: String,
        default: '${convertPxToRem(size)}'
      },
      fill: {
        type: String,
        required: true
      }
    }
  }
`
}

const vueIconsIndex = iconList => `<script>
${fileHeader}
${iconList.map(vueIconComponent).join('')}
</script>`

module.exports.vueIconsIndex = vueIconsIndex
