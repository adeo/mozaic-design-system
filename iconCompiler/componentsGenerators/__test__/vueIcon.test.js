import vueIcon from '../vueIcon'

const iconList = [
  {
    data: `<svg></svg>`,
    fileName: 'Navigation_Arrow_Arrow--Top_16px.svg',
    outputPath: 'toto',
  },
  {
    data: `<svg></svg>`,
    fileName: 'Media_FullScreen_24px.svg',
    outputPath: 'toto',
  },
]

describe('vueIcon', () => {
  it('vueIcon create a file with a list of exported icons', () => {
    expect(vueIcon(iconList)).toMatchSnapshot()
  })
})
