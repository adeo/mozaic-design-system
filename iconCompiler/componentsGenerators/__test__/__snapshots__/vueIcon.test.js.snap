// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vueIcon return a console error and does not output icon when data is missing 1`] = `
"<script>
/*
DON'T EDIT THIS FILE
It was auto-generated by \\"iconCompiler\\" script
*/

const GenerateIcon = (size, viewBox, childs) => ({
  props: {
    size: {
      type: String,
      default: size,
    },
    fill: {
      type: String,
      default: 'currentColor',
    },
  },
  render: function (createElement) {
    return createElement(
      'svg',
      {
        attrs: {
          height: this.size,
          width: this.size,
          fill: this.fill,
          viewBox: viewBox
        },
      },
      childs.map(elem => createElement(...elem))
    );
  },
});



export default {}
</script>"
`;

exports[`vueIcon vueIcon create a file with a list of exported icons 1`] = `
"<script>
/*
DON'T EDIT THIS FILE
It was auto-generated by \\"iconCompiler\\" script
*/

const GenerateIcon = (size, viewBox, childs) => ({
  props: {
    size: {
      type: String,
      default: size,
    },
    fill: {
      type: String,
      default: 'currentColor',
    },
  },
  render: function (createElement) {
    return createElement(
      'svg',
      {
        attrs: {
          height: this.size,
          width: this.size,
          fill: this.fill,
          viewBox: viewBox
        },
      },
      childs.map(elem => createElement(...elem))
    );
  },
});



export const MIconArrowArrowTop16 = GenerateIcon('1rem', '0 0 16 16', [
  [
    'path',
    { attrs: {\\"d\\":\\"M42 7H6a2 2 0 00-2 2v27a2 2 0 002 2h10v3h-4.5a1.5 1.5 0 000 3h25a1.5 1.5 0 000-3H32v-3h10a2 2 0 002-2V9a2 2 0 00-2-2zm-1 3v18H7V10zM30 35H7v-5h34v5zm0 6H18v-3h12z\\"} },
  ]
])

export const MIconFullScreen24 = GenerateIcon('1.5rem', '0 0 24 24', [
  [
    'path',
    { attrs: {\\"d\\":\\"M42 7H6a2 2 0 00-2 2v27a2 2 0 002 2h10v3h-4.5a1.5 1.5 0 000 3h25a1.5 1.5 0 000-3H32v-3h10a2 2 0 002-2V9a2 2 0 00-2-2zm-1 3v18H7V10zM30 35H7v-5h34v5zm0 6H18v-3h12z\\"} },
  ]
])

export default {}
</script>"
`;
